src/torch/README.md

```diff
@@ -0,0 +1,101 @@
+3.  `sirf.torch.ObjectiveFunctionGradient`: Wraps a SIRF `ObjectiveFunction` that computes the objective function gradients in the forward pass and the Hessian-vector product in the backward pass. In the backward the Hessian is evaluated at the point which the objective function's gradient was evaluated.
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the "upstream gradient" (`grad_output`).
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Applies the SIRF `Operator.backward()` method. This will apply the **Jacobian adjoint** of the operator to upstream gradient (the vector).
+    4.  Converts the resulting SIRF object back to a PyTorch tensor and returns it.
+
+### `_ObjectiveFunction` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor (representing an image for instance) to a SIRF object.
+    2.  Calls the SIRF `ObjectiveFunction.__call__()` method.
+    3.  Returns the of the objective function value as a PyTorch tensor.
+    4.  Saves relevant information to the `ctx` if gradients are needed.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), in this case it is always a scalar.
+    2.  Gets the gradient of the objective function using `sirf_obj_func.gradient()`, which computed at the input and multiplied by the upstream gradient.
+    3.  Converts the SIRF gradient to a PyTorch tensor.
+    4.  Returns the gradient multiplied by `grad_output`.
+
+
+### `_ObjectiveFunctionGradient` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Computes the *gradient* of the SIRF objective function using `sirf_obj_func.gradient()`, which is computed on the input.
+    3.  Returns the gradient as a PyTorch tensor.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), which now represents a *vector* (not a scalar) of the same shape as the output of `forward`.
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Multiples the Hessian evaluated at the input of `forward` with the "upstream gradient" using `sirf_obj_func.multiply_with_Hessian()`.
+    4.  Returns the Hessian multiplied with a vector as a tensor.
+
+# TODO
+
+* Extend to subsets in the wrapper
```

Comment:
Is this a use-case?

Reply:
at present we don't use subsets in the wrapper, nor do we have multiple objective functions